### ISSUE TITLE
Inference result is incorrect

### ISSUE THREAD

**Issue by chenlin038** [NVIDIA/DeepLearningExamples#436]
Related to **Model/Framework(s)** 
FasterTransformer v2

**Describe the bug**
I follow the instructions in the sample file(v2/sample/tensorflow_bert/sample.md).  Everything looks normal, but the inference result is incorrect.

**To Reproduce**
①Clone the repository.
git clone https://github.com/NVIDIA/DeepLearningExamples
cd DeepLearningExamples/FasterTransformer/v2
git submodule init
git submodule update
②Build the project.
ln -s /usr/lib/python2.7/site-packages/tensorflow/libtensorflow_framework.so.1 /usr/lib/python2.7/site-packages/tensorflow/libtensorflow_framework.so
mkdir -p build
cd build
cmake -DSM=70 -DCMAKE_BUILD_TYPE=Release -DBUILD_TF=ON -DTF_PATH=/usr/lib/python2.7/site-packages/tensorflow/ ..
③Generate the gemm_config.in file:
./bin/encoder_gemm 1 300 8 48 0
④Follow the instructions in the sample file: 
I used my own data to run run_classifier_wrap.py.There are no errors but the inference result is incorrect. After debugging, I found a problem with the "transformer_op_module" op.
**Firstly, If I don't use this op, then the inference result will be correct**. Further, I added print information before and after this op:
 # FASTINFER: fast transformer encoder inference
            trainable_vars = tf.get_collection(
                tf.GraphKeys.TRAINABLE_VARIABLES, scope=tf.get_variable_scope().name)
            **layer_input = tf.Print(layer_input, [layer_input, seq_length, num_attention_heads, attention_head_size], message='Debug message:layer_Input')**
            layer_output = transformer_op_module.bert_transformer(
                layer_input,
                layer_input,
                trainable_vars[0], trainable_vars[2], trainable_vars[4], trainable_vars[1], trainable_vars[3], trainable_vars[5],
                attention_mask,
                trainable_vars[6], trainable_vars[7], trainable_vars[8], trainable_vars[9], trainable_vars[10], trainable_vars[11],
                trainable_vars[12], trainable_vars[13], trainable_vars[14], trainable_vars[15],
                from_seq_len=seq_length, to_seq_len=seq_length, head_num=num_attention_heads, size_per_head=attention_head_size)
            **layer_output = tf.Print(layer_output, [layer_output], message='Debug message:layer_Output')**
Then I got:
Debug message:layer_Input[[0.170629382 0.539957643 -0.102443084...]...][300][8][48]
Debug message:layer_Output[[0 0 0...]...]
Debug message:layer_Input[[0 0 0...]...][300][8][48]
Debug message:layer_Output[[0 0 0...]...]
Debug message:layer_Input[[0 0 0...]...][300][8][48]
Debug message:layer_Output[[0 0 0...]...]
Debug message:layer_Input[[0 0 0...]...][300][8][48]
Debug message:layer_Output[[0 0 0...]...]
Debug message:layer_Input[[0 0 0...]...][300][8][48]
Debug message:layer_Output[[0 0 0...]...]
Debug message:layer_Input[[0 0 0...]...][300][8][48]
Debug message:layer_Output[[0 0 0...]...]
Debug message:layer_Input[[0 0 0...]...][300][8][48]
Debug message:layer_Output[[0 0 0...]...]
Debug message:layer_Input[[0 0 0...]...][300][8][48]
Debug message:layer_Output[[0 0 0...]...]

All values of layer_Output become 0, so is there any problem with this op?

**Expected behavior**
Inference results are consistent with no faster transormer used.

**Environment**
* Container version: **No container used.**
* GPUs in the system: TitanV
* CUDA driver version:418.67
* CUDA version: 10.1
* Tensorflow version: 1.14 python 2.7
* Cmake version:3.9.2

**Comment by byshiue** [comment]
@chenlin038 Can you run the encoder_sample.py correctly?

**Comment by chenlin038** [comment]
@byshiue 
after i run "./bin/encoder_gemm 1 300 8 48 0",  i run 
python encoder_sample.py \
   --batch_size 1 \
   --seq_len 300 \
   --head_number 8 \
   --size_per_head 48 \
   --num_layer 8 \
   --data_type fp32 \
   --test_time 1

i got:
#################################
2020-03-30 17:53:37.100322: I tensorflow/stream_executor/cuda/ptxas_utils.cc:202] 
2020-03-30 17:53:37.126336: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0
[INFO] Encoder Cross check False
[INFO] Max diff 4.47700500488
[INFO] min diff 2.21217051148e-05
[INFO] TF decoder time costs: 6.44892 ms
[INFO] OP decoder time costs: 4.06816 ms

**Comment by chenlin038** [comment]
@byshiue An important thing is that this time I used NGC container TensorFlow 19.10-py2.
And I completely followed the instructions to install FasterTransformer(FT) on my TitanV machine.
There are some results:
①Generate the gemm_config.in file:
./bin/encoder_gemm 1 300 8 48 0
②python encoder_sample.py 
--batch_size 1 
--seq_len 300 
--head_number 8 
--size_per_head 48 
--num_layer 8 
--data_type fp32 
--test_time 1
i got:
#################################
Successfully opened dynamic library libcublas.so.10.0
[INFO] Encoder Cross check **True**
[INFO] Max diff 4.76837158203e-06
[INFO] min diff 0.0
[INFO] TF decoder time costs: 3.86098 ms
[INFO] OP decoder time costs: 3.18552 ms

However, the problem of inconsistent inference still exists, and I once again located the problem caused by the "transformer_op_module".
I compared the original bert result with the FT  by tf.Print():
        trainable_vars = tf.get_collection(
            tf.GraphKeys.TRAINABLE_VARIABLES, scope=tf.get_variable_scope().name)
        **layer_input = tf.Print(layer_input, [layer_input, seq_length, num_attention_heads, attention_head_size], message='Debug message:layer_Input')**
        layer_output = transformer_op_module.bert_transformer(
            layer_input,
            layer_input,
            trainable_vars[0], trainable_vars[2], trainable_vars[4], trainable_vars[1], trainable_vars[3], trainable_vars[5],
            attention_mask,
            trainable_vars[6], trainable_vars[7], trainable_vars[8], trainable_vars[9], trainable_vars[10], trainable_vars[11],
            trainable_vars[12], trainable_vars[13], trainable_vars[14], trainable_vars[15],
            from_seq_len=seq_length, to_seq_len=seq_length, head_num=num_attention_heads, size_per_head=attention_head_size)
        **layer_output = tf.Print(layer_output, [layer_output], message='Debug message:layer_Output')**

And original bert result is:
Debug message:layer_Input[[0.170629382 0.539957643 -0.102443084...]...]
**Debug message:layer_Output[[0.145671055 0.308657706 -0.145599499...]...]**
Debug message:layer_Input[[0.145671055 0.308657706 -0.145599499...]...]
Debug message:layer_Output[[0.27708298 0.43354845 0.0147406515...]...]
Debug message:layer_Input[[0.27708298 0.43354845 0.0147406515...]...]
Debug message:layer_Output[[0.119168065 0.384940326 0.133568972...]...]
Debug message:layer_Input[[0.119168065 0.384940326 0.133568972...]...]
Debug message:layer_Output[[0.223199576 0.319623232 0.109987721...]...]
Debug message:layer_Input[[0.223199576 0.319623232 0.109987721...]...]
Debug message:layer_Output[[0.734694 1.24279058 -0.0975921527...]...]
Debug message:layer_Input[[0.734694 1.24279058 -0.0975921527...]...]
Debug message:layer_Output[[0.519900262 1.64103115 0.194583863...]...]
Debug message:layer_Input[[0.519900262 1.64103115 0.194583863...]...]
Debug message:layer_Output[[0.891126931 1.28525829 0.263341427...]...]
Debug message:layer_Input[[0.891126931 1.28525829 0.263341427...]...]
Debug message:layer_Output[[1.37031257 1.70470738 0.838114262...]...]

But Ft result is:
Debug message:layer_Input[[0.170629382 0.539957643 -0.102443084...]...]
**Debug message:layer_Output[[0.636132419 -0.694242358 -0.43078047...]...]**(It's starting to look different here)
Debug message:layer_Input[[0.636132419 -0.694242358 -0.43078047...]...]
Debug message:layer_Output[[0.753863811 0.120602824 0.253130615...]...]
Debug message:layer_Input[[0.753863811 0.120602824 0.253130615...]...]
Debug message:layer_Output[[0.327186853 1.04851806 -0.328607559...]...]
Debug message:layer_Input[[0.327186853 1.04851806 -0.328607559...]...]
Debug message:layer_Output[[-0.133049905 0.170171171 0.228999883...]...]
Debug message:layer_Input[[-0.133049905 0.170171171 0.228999883...]...]
Debug message:layer_Output[[0.551192045 1.15826392 0.0696898401...]...]
Debug message:layer_Input[[0.551192045 1.15826392 0.0696898401...]...]
Debug message:layer_Output[[1.62492692 0.323082924 0.368612677...]...]
Debug message:layer_Input[[1.62492692 0.323082924 0.368612677...]...]
Debug message:layer_Output[[-0.399213016 -0.0504479334 -0.655377507...]...]
Debug message:layer_Input[[-0.399213016 -0.0504479334 -0.655377507...]...]
Debug message:layer_Output[[-1.55534577 0.627369 0.668132126...]...]

So you can see that the result after transformer_op_module will be different.
This problem has troubled me for a long time and I really hope you can help me.

**Comment by byshiue** [comment]
> @byshiue An important thing is that this time I used NGC container TensorFlow 19.10-py2.
> And I completely followed the instructions to install FasterTransformer(FT) on my TitanV machine.
> There are some results:
> ①Generate the gemm_config.in file:
> ./bin/encoder_gemm 1 300 8 48 0
> ②python encoder_sample.py
> --batch_size 1
> --seq_len 300
> --head_number 8
> --size_per_head 48
> --num_layer 8
> --data_type fp32
> --test_time 1
> i got:
> #################################
> Successfully opened dynamic library libcublas.so.10.0
> [INFO] Encoder Cross check **True**
> [INFO] Max diff 4.76837158203e-06
> [INFO] min diff 0.0
> [INFO] TF decoder time costs: 3.86098 ms
> [INFO] OP decoder time costs: 3.18552 ms
> 
> However, the problem of inconsistent inference still exists, and I once again located the problem caused by the "transformer_op_module".
> I compared the original bert result with the FT by tf.Print():
> trainable_vars = tf.get_collection(
> tf.GraphKeys.TRAINABLE_VARIABLES, scope=tf.get_variable_scope().name)
> **layer_input = tf.Print(layer_input, [layer_input, seq_length, num_attention_heads, attention_head_size], message='Debug message:layer_Input')**
> layer_output = transformer_op_module.bert_transformer(
> layer_input,
> layer_input,
> trainable_vars[0], trainable_vars[2], trainable_vars[4], trainable_vars[1], trainable_vars[3], trainable_vars[5],
> attention_mask,
> trainable_vars[6], trainable_vars[7], trainable_vars[8], trainable_vars[9], trainable_vars[10], trainable_vars[11],
> trainable_vars[12], trainable_vars[13], trainable_vars[14], trainable_vars[15],
> from_seq_len=seq_length, to_seq_len=seq_length, head_num=num_attention_heads, size_per_head=attention_head_size)
> **layer_output = tf.Print(layer_output, [layer_output], message='Debug message:layer_Output')**
> 
> And original bert result is:
> Debug message:layer_Input[[0.170629382 0.539957643 -0.102443084...]...]
> **Debug message:layer_Output[[0.145671055 0.308657706 -0.145599499...]...]**
> Debug message:layer_Input[[0.145671055 0.308657706 -0.145599499...]...]
> Debug message:layer_Output[[0.27708298 0.43354845 0.0147406515...]...]
> Debug message:layer_Input[[0.27708298 0.43354845 0.0147406515...]...]
> Debug message:layer_Output[[0.119168065 0.384940326 0.133568972...]...]
> Debug message:layer_Input[[0.119168065 0.384940326 0.133568972...]...]
> Debug message:layer_Output[[0.223199576 0.319623232 0.109987721...]...]
> Debug message:layer_Input[[0.223199576 0.319623232 0.109987721...]...]
> Debug message:layer_Output[[0.734694 1.24279058 -0.0975921527...]...]
> Debug message:layer_Input[[0.734694 1.24279058 -0.0975921527...]...]
> Debug message:layer_Output[[0.519900262 1.64103115 0.194583863...]...]
> Debug message:layer_Input[[0.519900262 1.64103115 0.194583863...]...]
> Debug message:layer_Output[[0.891126931 1.28525829 0.263341427...]...]
> Debug message:layer_Input[[0.891126931 1.28525829 0.263341427...]...]
> Debug message:layer_Output[[1.37031257 1.70470738 0.838114262...]...]
> 
> But Ft result is:
> Debug message:layer_Input[[0.170629382 0.539957643 -0.102443084...]...]
> **Debug message:layer_Output[[0.636132419 -0.694242358 -0.43078047...]...]**(It's starting to look different here)
> Debug message:layer_Input[[0.636132419 -0.694242358 -0.43078047...]...]
> Debug message:layer_Output[[0.753863811 0.120602824 0.253130615...]...]
> Debug message:layer_Input[[0.753863811 0.120602824 0.253130615...]...]
> Debug message:layer_Output[[0.327186853 1.04851806 -0.328607559...]...]
> Debug message:layer_Input[[0.327186853 1.04851806 -0.328607559...]...]
> Debug message:layer_Output[[-0.133049905 0.170171171 0.228999883...]...]
> Debug message:layer_Input[[-0.133049905 0.170171171 0.228999883...]...]
> Debug message:layer_Output[[0.551192045 1.15826392 0.0696898401...]...]
> Debug message:layer_Input[[0.551192045 1.15826392 0.0696898401...]...]
> Debug message:layer_Output[[1.62492692 0.323082924 0.368612677...]...]
> Debug message:layer_Input[[1.62492692 0.323082924 0.368612677...]...]
> Debug message:layer_Output[[-0.399213016 -0.0504479334 -0.655377507...]...]
> Debug message:layer_Input[[-0.399213016 -0.0504479334 -0.655377507...]...]
> Debug message:layer_Output[[-1.55534577 0.627369 0.668132126...]...]
> 
> So you can see that the result after transformer_op_module will be different.
> This problem has troubled me for a long time and I really hope you can help me.

Do you mean that the FT can return same results to TensorFlow in the encoder_sample.py, but fail for your test codes? 

Do you run the FT and TensorFlow results in one session run in your test codes?

**Comment by chenlin038** [comment]
@byshiue I show the result of encoder_sample.py just to show that the FT installation is successfulI. The function of encoder_sample.py is to verify whether the FT is successfully installed, isn't it?And I run the FT and TensorFlow results in one session run in my test codes.
Moreover, I have sent details to your email bhsueh@nvidia.com.

**Comment by byshiue** [comment]
The problem is caused by the intermediate hidden dimension of FFN are not hidden dim * 4 in this case.

**Comment by chenlin038** [comment]
@byshiue Thank you for your help!
